### PR TITLE
171 implement multi table helper functions (second round)

### PR DESCRIPTION
**Pull request by folmos-at-orange**

### CODE REVIEW

File: doc/samples/samples.rst [hunk 1]
```diff
@@ -1185,6 +1185,34 @@ Samples
         output_data_table_path,
         sort_variables=["AccidentId", "VehicleId"],
     )
+.. autofunction:: sort_data_tables_mt
```
Comment (popescu-v) [inline doc/samples/samples.rst:1188]: AFAIR, we should only have the `train_test_split` multi-table here, not the `sort`. Hence, this should be removed from the current PR and added to a different one I guess.

File: doc/samples/samples_sklearn.rst [hunk 2]
```diff
@@ -365,20 +417,9 @@ Samples
...
-    # Load the dataset into a pandas dataframe
-    iris_path = os.path.join(kh.get_samples_dir(), "Iris", "Iris.txt")
-    iris_df = pd.read_csv(iris_path, sep="\t")
-
-    # Train the model with the whole dataset
-    X = iris_df.drop(["Class"], axis=1)
-    y = iris_df["Class"]
-    khc = KhiopsClassifier()
-    khc.fit(X, y)
+    from sklearn.datasets import load_iris
```
Comment (popescu-v) [inline doc/samples/samples_sklearn.rst:422]: This goes beyond the scope of the PR IMHO.

File: doc/samples/samples_sklearn.rst [hunk 3]
```diff
@@ -459,29 +497,22 @@ Samples
...
 
-    # Load the dataset into a pandas dataframe
-    iris_path = os.path.join(kh.get_samples_dir(), "Iris", "Iris.txt")
-    iris_df = pd.read_csv(iris_path, sep="\t")
-
-    # Train the model with the whole dataset
-    X = iris_df.drop("Class", axis=1)
-    y = iris_df["Class"]
+    # Load the dataset
+    X, y = load_iris(return_X_y=True)
```
Comment (popescu-v) [inline doc/samples/samples_sklearn.rst:504]: IMHO this helper should be in a different PR.

File: khiops/core/internals/runner.py [hunk 4]
```diff
@@ -78,7 +78,7 @@ def get_dir_status(a_dir):
     return status
 
 
-def check_samples_dir(samples_dir):
+def _check_samples_dir(samples_dir):
```
Comment (popescu-v) [inline khiops/core/internals/runner.py:81]: This should disappear after rebase on `dev`.

File: khiops/samples/samples.py [hunk 5]
```diff
@@ -1329,6 +1329,35 @@ def sort_data_table_expert():
     )
 
 
+def sort_data_tables_mt():
```
Comment (popescu-v) [inline khiops/samples/samples.py:1332]: This should be in a different PR (the scope of the current PR is to only add the `train_test_split` multi-table).

File: khiops/samples/samples.py [hunk 6]
```diff
@@ -1662,6 +1691,7 @@ def build_deployed_dictionary():
     deploy_regressor_for_metrics,
     sort_data_table,
     sort_data_table_expert,
+    sort_data_tables_mt,
```
Comment (popescu-v) [inline khiops/samples/samples.py:1694]: Out of the scope of the current PR AFAIU.

File: khiops/samples/samples_sklearn.py [hunk 7]
```diff
@@ -390,9 +428,14 @@ def khiops_classifier_pickle():
     else:
         os.makedirs(results_dir, exist_ok=True)
 
+    # Train the model with the Iris dataset
+    X, y = load_iris(return_X_y=True)
```
Comment (popescu-v) [inline khiops/samples/samples_sklearn.py:432]: This helper seems of the scope of the current PR AFAIU.

File: khiops/samples/samples_sklearn.py [hunk 8]
```diff
@@ -472,29 +509,22 @@ def khiops_encoder():
...
-
-    # Train the model with the whole dataset
-    X = iris_df.drop("Class", axis=1)
-    y = iris_df["Class"]
+    # Load the dataset
+    X, y = load_iris(return_X_y=True)
 
     # Create the encoder object
-    khe = KhiopsEncoder()
+    khe = KhiopsEncoder(transform_type_numerical="part_label")
```
Comment (popescu-v) [inline khiops/samples/samples_sklearn.py:519]: Why is this change needed? Could you add a comment to explain?

File: khiops/sklearn/estimators.py [hunk 9]
```diff
@@ -86,70 +91,110 @@ def _extract_basic_dictionary(dictionary):
 
 def _check_dictionary_compatibility(
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:92]: Why are these changes part of the scope of the current PR? It seems to me they are not. But if they are, they should occur in a separate, clearly-identified commit IMHO, which states the rationale of these changes being included in this PR.

File: khiops/sklearn/estimators.py [hunk 10]
```diff
@@ -495,15 +539,15 @@ def _transform_deploy_model(
         # Set output path files
         output_dir = self._get_output_dir(computation_dir)
         log_file_path = fs.get_child_path(output_dir, "khiops.log")
-        output_data_table_path = fs.get_child_path(output_dir, "transformed.txt")
+        output_data_table_path = fs.get_child_path(output_dir, transformed_file_name)
 
         # Set the format parameters depending on the type of dataset
-        if deployment_dataset.is_in_memory():
+        if deployment_ds.is_in_memory:
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:545]: This change is out of the scope of the current PR AFAIU. If ancillary improvements are included in the PR, then they should be a in a separate commit or set of commits clearly identifying these changes as such and explaining the rationale for them.

File: khiops/sklearn/estimators.py [hunk 11]
```diff
@@ -1162,51 +1206,49 @@ def predict(self, X):
...
         # Call the parent method
-        super()._transform_check_dataset(dataset)
+        super()._transform_check_dataset(ds)
 
         # Coclustering models are special:
         # - They are mono-table only
         # - They are deployed with a multitable model whose main table contain
         #   the keys of the input table and the secondary table is the input table
-        if dataset.is_multitable():
+        if ds.is_multitable:
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:1251]: Idem, this seems out of the scope of the PR. It should be included in a separate commit / set of commits stating the rationale for it.

File: khiops/sklearn/estimators.py [hunk 12]
```diff
@@ -1584,29 +1616,29 @@ def _fit_training_post_process(self, dataset):
...
             )
 
         # Error if different number of dictionaries
-        dataset_domain = dataset.create_khiops_dictionary_domain()
-        if len(self.model_.dictionaries) != len(dataset_domain.dictionaries):
+        ds_domain = ds.create_khiops_dictionary_domain()
+        if len(self.model_.dictionaries) != len(ds_domain.dictionaries):
             raise ValueError(
-                f"X has {len(dataset_domain.dictionaries)} table(s), "
+                f"X has {len(ds_domain.dictionaries)} table(s), "
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:1641]: All these variable renamings should be in a separate commit for better "reviewability" of the PR (although they don't seem to be necessary to the PR).

File: khiops/sklearn/estimators.py [hunk 13]
```diff
@@ -1707,10 +1741,13 @@ def predict(self, X):
         assert isinstance(y_pred, (str, pd.DataFrame)), "Expected str or DataFrame"
         return y_pred
 
-    def _transform_prepare_deployment_model_for_predict(self):
+    def _transform_prepare_deployment_model_for_predict(self, ds):
         assert (
             self._predicted_target_meta_data_tag is not None
         ), "Predicted target metadata tag is not set"
+        assert hasattr(
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:1748]: Why is this refactoring needed in this PR? If it is needed, then it should go in a separate commit (different from the commit that introduces the `train_test_split_dataset` helper) that explains the rationale.

File: doc/samples/samples_sklearn.rst (hunk 2 above)
Comment (folmos-at-orange) [reply doc/samples/samples_sklearn.rst:422]: Ok

File: khiops/core/internals/runner.py (hunk 4 above)
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py]: It did.

File: khiops/core/internals/runner.py [hunk 14]
```diff
@@ -306,20 +306,15 @@ def _check_executable(bin_path):
 
 
 def get_linux_distribution_name():
-    """Detect Linux distribution name
+    """Detect the Linux distribution name
```
[the remark this answers is not in the document]
Comment (folmos-at-orange) [reply khiops/core/internals/runner.py]: It did

File: khiops/samples/samples_sklearn.py (hunk 7 above)
Comment (folmos-at-orange) [reply khiops/samples/samples_sklearn.py:432]: Ok

File: khiops/sklearn/estimators.py (hunk 9 above)
Comment (folmos-at-orange) [reply khiops/sklearn/estimators.py:92]: This change is ok after offline discussion.

File: khiops/sklearn/estimators.py (hunk 10 above)
Comment (folmos-at-orange) [reply khiops/sklearn/estimators.py:545]: This change is ok after offline discussion.

File: khiops/sklearn/estimators.py (hunk 11 above)
Comment (folmos-at-orange) [reply khiops/sklearn/estimators.py:1251]: This change is ok after offline discussion.

File: khiops/utils/helpers.py [hunk 15]
```diff
@@ -135,52 +41,43 @@ def train_test_split_dataset(
...
-        sklearn_split_params = {}
-        for param in ("train_size", "random_state", "shuffle", "stratify"):
-            if param in kwargs:
-                sklearn_split_params[param] = kwargs[param]
-
-        if target_column is None:
-            train_ds, test_ds = _train_test_split_in_memory_dataset(
+    # Check the table type of the dataset
+    if isinstance(ds.main_table, FileTable):
+        raise ValueError("Only Pandas and Numpy datasets may be used in this method.")
```
Comment (popescu-v) [inline khiops/utils/helpers.py:46]: I would slightly rephrase to "Only Pandas dataframe or NumPy array-based datasets can be used in this method."

Comment (popescu-v) [reply khiops/utils/helpers.py:46]: Also, this also works with sparse datasets, right? If so, then we should mention it, right?

File: khiops/utils/helpers.py [hunk 16]
```diff
@@ -135,52 +41,43 @@ def train_test_split_dataset(
...
+
+    # Obtain the keys for the other test_train_split function
+    sklearn_split_params = {}
+    for param in ("train_size", "random_state", "shuffle", "stratify"):
+        if param in kwargs:
+            sklearn_split_params[param] = kwargs[param]
+
+    # Perform the split with and without target
+    if y is None:
+        train_ds, test_ds = _train_test_split_in_memory_dataset(
```
Comment (popescu-v) [inline khiops/utils/helpers.py:56]: IMHO, `_train_test_split_in_memory_dataset` should have the same return type (3-tuple) in all cases. If `y` is `None`, then the returned target could be also `None`.

Thus, the `if/else` block can be simplified: only `train_target_column = None` would remain in the `if y is None` block.

File: khiops/samples/samples_sklearn.py (hunk 8 above)
Comment (popescu-v) [reply khiops/samples/samples_sklearn.py:519]: Isn't it `"part label"` as per the documentation (w/o underscore)?

File: khiops/samples/samples_sklearn.py [hunk 17]
```diff
@@ -219,42 +198,106 @@ def khiops_classifier_multitable_star():
...
+            "Vehicles": (
+                os.path.join(accidents_data_dir, "Vehicles.txt"),
+                ["AccidentId", "VehicleId"],
+            ),
+        },
+        "format": ("\t", True),
+    }
+
+    # Split the dataset into train and test
+    X_train, X_test = train_test_split_dataset(
```
Comment (popescu-v) [inline khiops/samples/samples_sklearn.py:233]: `train_test_split_dataset` does not support file datasets. Hence, this sample should not be updated to using the new train/test split feature. Otherwise, the sample fails.

File: khiops/utils/helpers.py [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+# unbalanced-tuple-unpacking. See issue https://github.com/pylint-dev/pylint/issues/5671
+
+
+def train_test_split_dataset(ds_spec, y=None, test_size=0.25, **kwargs):
+    """Splits a dataset spec into train and test
+
+    Parameters
+    ----------
+    ds_spec : ``dict``
+        A dataset spec. The tables must be either `pandas.DataFrame` or file path
```
Comment (popescu-v) [inline khiops/utils/helpers.py:20]: Remove `"or file path references"` (these are not supported).

**Review by popescu-v — CHANGES_REQUESTED** [review]
See comments.

File: khiops/utils/helpers.py (hunk 15 above)
Comment (folmos-at-orange) [reply khiops/utils/helpers.py:46]: Ok

File: khiops/samples/samples_sklearn.py (hunk 17 above)
Comment (folmos-at-orange) [reply khiops/samples/samples_sklearn.py]: Ok for remove & Done

File: khiops/samples/samples_sklearn.py (hunk 8 above)
Comment (folmos-at-orange) [reply khiops/samples/samples_sklearn.py:451]: Yes, the encoded variables are the intervals, so it is more pedagogical.

File: khiops/samples/samples_sklearn.py (hunk 7 above)
Comment (folmos-at-orange) [reply khiops/samples/samples_sklearn.py:364]: Done

File: doc/samples/samples_sklearn.rst (hunk 3 above)
Comment (folmos-at-orange) [reply doc/samples/samples_sklearn.rst:437]: Done

File: doc/samples/samples_sklearn.rst (hunk 2 above)
Comment (folmos-at-orange) [reply doc/samples/samples_sklearn.rst:355]: Reversion Done

File: khiops/utils/helpers.py (hunk 16 above)
Comment (folmos-at-orange) [reply khiops/utils/helpers.py:56]: Ok & Done.

I left the 2-tuple or 4-tuple pattern only the return of the `train_test_split_dataset` so it matches that of sklearn.

File: khiops/utils/helpers.py (hunk 15 above)
Comment (folmos-at-orange) [reply khiops/utils/helpers.py:46]: Added `Scipy sparse`, Done.

File: khiops/samples/samples_sklearn.py (hunk 8 above)
Comment (popescu-v) [reply khiops/samples/samples_sklearn.py:459]: Ok, fair enough.

File: khiops/utils/helpers.py (hunk 16 above)
Comment (popescu-v) [reply khiops/utils/helpers.py]: Yes, ok for keeping the 2-/4- tuple pattern for the public helper.

File: khiops/utils/helpers.py (hunk 18 above)
Comment (folmos-at-orange) [reply khiops/utils/helpers.py]: Ok.

I uniformized with the docstring of estimators, added the file-path caveat and added a `Raises` section.

File: khiops/utils/helpers.py [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+
+
+def train_test_split_dataset(ds_spec, y=None, test_size=0.25, **kwargs):
+    """Splits a dataset spec into train and test
+
+    Parameters
+    ----------
+    ds_spec : ``dict``
+         A ``dict`` multi-table dataset specification (see :doc:`/multi_table_primer`).
+         File-path based datasets are not accepted.
```
Comment (popescu-v) [inline khiops/utils/helpers.py:21]: I would avoid negative phrasings, and rather eliminate this statement altogether, as file-path datasets are deprecated and would go away anyway. Instead, I would add a positive formulation that would match the error message: `Pandas, NumPy, SciPy sparse datasets are accepted.`

File: khiops/utils/helpers.py (hunk 18 above)
Comment (popescu-v) [reply khiops/utils/helpers.py]: OK, but see the other new comment on this. Closing this thread, as the new thread would remain.

File: khiops/utils/helpers.py (hunk 19 above)
Comment (folmos-at-orange) [reply khiops/utils/helpers.py]: Done rephrased.

Comment (popescu-v) [reply khiops/utils/helpers.py]: Ok, with the caveat to remove the `Only` once file datasets are removed (for v11).

**Review by popescu-v — APPROVED** [review]
LGTM.
